### PR TITLE
Speed up coordinate representations using erfa routines

### PR DESCRIPTION
**Pull request by mhvk**
Follow-up of #7638, using `erfa` ufuncs for conversion of cartesian to (unit)spherical, and some of the cartesian methods.

Sample speed-up:
```
import numpy as np
from astropy.coordinates import CartesianRepresentation, SphericalRepresentation
cr = CartesianRepresentation(np.ones((100, 3)), unit=u.km, xyz_axis=-1)
%timeit cr.represent_as(SphericalRepresentation)
# 1000 loops, best of 3: 200 -> 174 µs
sph = cr.represent_as(SphericalRepresentation)
%timeit sph.represent_as(CartesianRepresentation)
# 1000 loops, best of 3: 233 -> 67 µs
%timeit cr.norm()
# 1000 loops, best of 3: 227 -> 9 µs
%timeit cr.dot(cr)
# 1000 loops, best of 3: 206 -> 28 µs
%timeit cr.cross(cr)
# 1000 loops, best of 3: 447 -> 66 µs
```

### CODE REVIEW

File: astropy/units/quantity.py [hunk 1]
```diff
@@ -714,7 +714,8 @@ def to_value(self, unit=None, equivalencies=[]):
                     # not in-place!
                     value = value * scale
 
-        return value if self.shape else value.item()
+        return value if self.shape else (value[()] if self.dtype.fields
```
Comment (mhvk) [inline astropy/units/quantity.py]: Probably this should be unconditionally `value[()]` instead of `value.item()` but out of scope of this PR...

**Comment by adrn** [comment]
I played around with this and did some verification that the ERFA functions are indeed very fast (compared even to numpy). 

One thing I don't like about this PR is that, in some cases, it hides some fairly simple Python code into obscurely-named ERFA functions that makes it much less readable. That's mainly just an observation and something I'd like to discuss, rather than something that needs to be fixed in the PR, but I'm interested to hear your thoughts on solutions for readability. Maybe we should add better comments to the code that explain the ERFA functions (as you do in some places)?

**Comment by mhvk** [comment]
@adrn - I agree with the part that this makes the code less readable, and am even fairly sure I commented myself pointing that out on an earlier suggestion to wrap these functions in C! (not #5735, not #4664, even though that is being "fixed" here... just cannot seem to find it!).

My current sense is that using anything from `erfa` inside astropy is fair game, just as I wouldn't hesitate to use a possibly obscure numpy function. That said, erfa functions are arguably harder to look up, especially since we keep them in a private module, so adding comments is reasonable.

Also cc @eteq, given #4664

**Comment by mhvk** [comment]
@adrn - since I had to rebase anyway, I went ahead and added simple one-line comments for each use of an erfa ufunc (basically, just taking the first line of their docstrings).

**Comment by astrofrog** [comment]
This looks nice! Though @eteq and/or @adrn should review/approve

**Comment by astrofrog** [comment]
What we could consider doing in future is just create aliases for the ERFA functions with better names

**Comment by mhvk** [comment]
> What we could consider doing in future is just create aliases for the ERFA functions with better names

I thought about it, but in the end felt it would just be confusing (`Time` used to do that). I think adding comments was the right thing to do - this will be even more important to do inside `builtin_frames` - things like `atoq` are truly obscure!

**Pull request merged**